File: AB_Testing/nodejs/v1.js
const COOKIE_EXPERIMENT_A = 'X-Experiment-Name=A';
const COOKIE_EXPERIMENT_B = 'X-Experiment-Name=B';
const PATH_EXPERIMENT_A = '/experiment-A';
const PATH_EXPERIMENT_B = '/experiment-B';

async function f(event) {
    let request = event.request;

    // 非测试页直接回源
    if (request.uri !== '/experiment') {
        return request;
    }

    let headers = request.headers;
    let experimentUri;

    // 遵循 nodejs 默认设定
    // http://nodejs.cn/api/http.html#http_class_http_incomingmessage
    if (typeof headers.cookie === 'string') {
        if (headers.cookie.indexOf(COOKIE_EXPERIMENT_A) >= 0) {
            experimentUri = PATH_EXPERIMENT_A;
        } else if (headers.cookie.indexOf(COOKIE_EXPERIMENT_B) >= 0) {
            experimentUri = PATH_EXPERIMENT_B;
        }
    }

    // 没有 cookie 则随机选取
    if (!experimentUri) {
        if (Math.random() < 0.75) {
            experimentUri = PATH_EXPERIMENT_A;
        } else {
            experimentUri = PATH_EXPERIMENT_B;
        }
    }

    request.uri = experimentUri;

    // runtime 自动回源
    return request;
}

exports.handler = f;

File: AB_Testing/python/v1.py
import random

COOKIE_EXPERIMENT_A = 'X-Experiment-Name=A'
COOKIE_EXPERIMENT_B = 'X-Experiment-Name=B'
PATH_EXPERIMENT_A = '/experiment-A'
PATH_EXPERIMENT_B = '/experiment-B'


async def handler(event):
    request = event.request

    # 非测试页直接回源
    if request.uri != '/experiment':
        return Request(request=request, config=event.config)

    headers = request.headers
    experiment_uri = None

    # headers 是 CIMultiDict
    # https://multidict.readthedocs.io/en/stable/multidict.html#cimultidict
    if 'cookie' in headers:
        cookies = headers.getall('cookie')

        for cookie in cookies:
            if cookie.find(COOKIE_EXPERIMENT_A) >= 0:
                experiment_uri = PATH_EXPERIMENT_A
                break
            elif cookie.find(COOKIE_EXPERIMENT_B) >= 0:
                experiment_uri = PATH_EXPERIMENT_B

    # 没有 cookie 则随机选取
    if experiment_uri is None:
        if random.random() < 0.75:
            experiment_uri = PATH_EXPERIMENT_A
        else:
            experiment_uri = PATH_EXPERIMENT_B

    request.uri = experiment_uri

    # runtime 自动回源
    return Request(request=request, config=event.config)